### PR TITLE
CMCL-1305: Confiner2D usability improvement

### PR DESCRIPTION
**Pull request by gaborkb**
### Purpose of this PR
[CMCL-1305](https://jira.unity3d.com/browse/CMCL-1305): 

- Info box message added when the camera window is bigger than what can fit inside the confiner and oversize window option is not enabled.
InfoBox

NoInfoBox

------------------------------------------------------------

- Tooltips clarified for the invalidate cache buttons clarified.
- Sample scene help message has more information about cache invalidation.

### Testing status

- [ ] Added an automated test
- [ ] Passed all automated tests
- [x] Manually tested 

### Documentation status

- [ ] Updated [CHANGELOG](https://keepachangelog.com/en/1.0.0/) - todo

### CODE REVIEW

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 1]
```diff
@@ -115,21 +118,21 @@ static void DrawConfinerGizmos(CinemachineConfiner2D confiner2D, GizmoType type)
                 return;
 
             var color = CinemachineCorePrefs.BoundaryObjectGizmoColour.Value;
-            var colorDimmed = new Color(color.r, color.g, color.b, color.a / 2f);
+            var colorDimmed = new Color(color.r, color.g, color.b, color.a / 0.5f);
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:121]: dividing by 0.5 is like multiplying b 2.  You are making it brighter, unless it's already 1 in which case you're doing nothing

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 2]
```diff
@@ -35,6 +35,10 @@ public override VisualElement CreateInspectorGUI()
             var volumeProp = serializedObject.FindProperty(() => Target.BoundingShape2D);
             ux.Add(new PropertyField(volumeProp));
             ux.Add(new PropertyField(serializedObject.FindProperty(() => Target.Damping)));
+            var confiner = Target; // so it gets captured in the lambdas
+            if (confiner.IsCameraOversizedForTheConfiner())
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:39]: This doesn't work.  With UITK, you have to check IsCameraOversizedForTheConfiner in an event callback, not at UI creation time.

**Review by glabute — CHANGES_REQUESTED** [review]

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs (hunk 2 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:39]: That's why it did not work properly. Thanks

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs (hunk 1 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:121]: That's a typo :) not even sure how it got changed.

The color is up for discussion. Julien will get back to me.

Comment (glabute) [reply com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:121]: I think the inner polygon should be dimmer

Comment (gaborkb) [reply com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:121]: Yes, that's how it was originally.

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 3]
```diff
@@ -114,22 +118,22 @@ static void DrawConfinerGizmos(CinemachineConfiner2D confiner2D, GizmoType type)
             if (!confiner2D.GetGizmoPaths(out var originalPath, ref s_CurrentPathCache, out var pathLocalToWorld))
                 return;
 
-            var color = CinemachineCorePrefs.BoundaryObjectGizmoColour.Value;
-            var colorDimmed = new Color(color.r, color.g, color.b, color.a / 2f);
-            
+            var pathCacheColor = CinemachineCorePrefs.BoundaryObjectGizmoColour.Value;
+            var inputColliderColor = new Color(145f / 255f, 244f / 255f, 139f / 255f, 1f);
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:122]: I don't like this hardcoded color.  It's better to make it a computed variant of the other color.  I still think the input polygon should be brighter than the computed one.  What was wrong with the colors before?

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 4]
```diff
@@ -35,8 +34,12 @@ public override VisualElement CreateInspectorGUI()
             var volumeProp = serializedObject.FindProperty(() => Target.BoundingShape2D);
             ux.Add(new PropertyField(volumeProp));
             ux.Add(new PropertyField(serializedObject.FindProperty(() => Target.Damping)));
+            var oversizedCameraHelp = ux.AddChild(new HelpBox(
+                "The camera window size is bigger than what can fit perfectly in the confiner. Enable Oversize Window option.",
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:38]: Awkward wording.  This is better:
"The camera window too big for the confiner. Enable the Oversize Window option."

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 5]
```diff
@@ -48,24 +51,27 @@ void TrackVolume(SerializedProperty p)
...
             {
+                oversizedCameraHelp.SetVisible(false);
                 if (confiner == null)
                     return; // target deleted
+                
                 if (!confiner.OversizeWindow.Enabled)
                 {
                     bakeTimeout.SetVisible(false);
                     bakeProgress.SetVisible(false);
+                    oversizedCameraHelp.SetVisible(confiner.IsCameraOversizedForTheConfiner());
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:74]: This is not appearing and disappearing for me when I change the lens size.  I have to slide something in the UX to trigger it (e.g. change the game window size).  Unity 2022.2f1.

**Review by glabute — CHANGES_REQUESTED** [review]

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs (hunk 5 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:74]: I can't reproduce the issue. I have to click Invalidate Lens Cache to make the text box appear/disappear.
Which makes me think:
We should check whether the Lens Cache is valid in the editor script. Let them know, with a warning, that the lens cache is not valid, they should invalidate it.

**Comment by JulienDidierjean** [comment]
All good on the UX side for the shorter Tooltips and the colors of the confiner (Green for the editable one and yellow for the computed one when selecting the camera, and green when selecting and adjusting the manual confiner when it's selected)   I didn't check the message in the Sample though

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 6]
```diff
@@ -84,7 +90,7 @@ void UpdateBakingProgress()
             { 
                 text = "Invalidate Lens Cache",
                 tooltip = "Invalidates the lens cache, so a new one is computed next frame.  " 
-                    + "Call this when when the Field of View, Orthographic Size, or aspect ratio changes."
+                    + "Call this when when the Lens of the camera changes."
```
Comment (JWalker0007) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:93]:
```suggestion
                    + "Call this when the Lens of the camera changes."
```

**Review by JWalker0007 — APPROVED** [review]

**Review by JulienDidierjean — APPROVED** [review]
all good for gizmos colors and text in tooltips being simplified and shortened

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 7]
```diff
@@ -35,7 +35,7 @@ public override VisualElement CreateInspectorGUI()
             ux.Add(new PropertyField(volumeProp));
             ux.Add(new PropertyField(serializedObject.FindProperty(() => Target.Damping)));
             var oversizedCameraHelp = ux.AddChild(new HelpBox(
-                "The camera window size is bigger than what can fit perfectly in the confiner. Enable Oversize Window option.",
+                "The camera window too big for the confiner. Enable the Oversize Window option.",
```
Comment (JWalker0007) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:38]:
```suggestion
                "The camera window is too big for the confiner. Enable the Oversize Window option.",
```

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 8]
```diff
@@ -33,48 +34,69 @@ public override VisualElement CreateInspectorGUI()
...
-            
-            ux.Add(new PropertyField(serializedObject.FindProperty(() => Target.OversizeWindow)));
-            
             TrackVolume(volumeProp);
             ux.TrackPropertyValue(volumeProp, TrackVolume);
             void TrackVolume(SerializedProperty p)
             {
                 var c = p.objectReferenceValue;
-                boundsHelp.SetVisible(!(c is PolygonCollider2D || c is BoxCollider2D || c is CompositeCollider2D));
+                boundsHelp.SetVisible(!(c is null || c is PolygonCollider2D || c is BoxCollider2D || c is CompositeCollider2D));
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:42]: Why this change?  Doesn't it make sense to have the help message if the shape is null also?

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 9]
```diff
@@ -33,48 +34,69 @@ public override VisualElement CreateInspectorGUI()
...
+            
+            var oversizedCameraHelp = ux.AddChild(new HelpBox(
+                "The camera window is too big for the confiner. Enable the Oversize Window option.",
+                HelpBoxMessageType.Info));
+
+            var owner = Target.VirtualCamera;
+            if (owner != null)
+            {
+                UpdateConfinerLensCache();
+                var lensProperty = new SerializedObject(owner).FindProperty("Lens");
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:59]: This is a nice thought, but it's also possible to add a confiner to a non-CmCamera - e.g. a mixing camera - and that doesn't have a "Lens" property.  So this code will fail in that case.

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs [hunk 10]
```diff
@@ -33,48 +34,69 @@ public override VisualElement CreateInspectorGUI()
...
+                oversizedCameraHelp.TrackPropertyValue(lensProperty, TrackLens);
+                void TrackLens(SerializedProperty lens) => UpdateConfinerLensCache();
+                void UpdateConfinerLensCache()
```
Comment (glabute) [inline com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:62]: Should this logic be in the runtime code, in case someone wants to do this at runtime?

**Review by glabute — CHANGES_REQUESTED** [review]

File: com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs (hunk 8 above)
Comment (gaborkb) [reply com.unity.cinemachine/Editor/Editors/CinemachineConfiner2DEditor.cs:42]: I felt no, but you are right, it makes sense to show it when null.

File: com.unity.cinemachine/Runtime/Behaviours/CinemachineConfiner2D.cs [hunk 11]
```diff
@@ -498,11 +511,70 @@ void CalculateDeltaTransformationMatrix()
...
+            static bool IsLensCacheInvalid(CameraState state, VcamExtraState extra, ShapeCache shapeCache)
+            {
+                var lens = state.Lens;
+                var deltaW = shapeCache.DeltaWorldToBaked;
+                var frustum = 
+                    CalculateHalfFrustumHeight(lens, deltaW.MultiplyPoint3x4(state.GetCorrectedPosition()).z);
+                return Mathf.Abs(extra.AspectRatio - lens.Aspect) > UnityVectorExtensions.Epsilon ||
+                    Mathf.Abs(extra.FrustumHeight - frustum * deltaW.lossyScale.x) > UnityVectorExtensions.Epsilon;
+            }
+        }
```
Comment (gaborkb) [inline com.unity.cinemachine/Runtime/Behaviours/CinemachineConfiner2D.cs:572]: Should this be public?

If so, should it follow the pattern of InvalidateLensCache (that it ignores ManagerCameras)?
Or should InvalidateLensCache follow this pattern, where we check whether we are dealing with manager cameras.

What do you think?

Comment (gaborkb) [reply com.unity.cinemachine/Runtime/Behaviours/CinemachineConfiner2D.cs:572]: https://github.com/Unity-Technologies/com.unity.cinemachine/pull/759 resolved question

**Review by glabute — APPROVED** [review]

**Comment by glabute** [comment]
Can't jump onto the platforms in Camera Magnets scene.  Player jump value needs tuning.
